libkbfs/mdserver_branch_journal.go

```diff
@@ -0,0 +1,159 @@
+}
+
+func ordinalToRevision(o journalOrdinal) (MetadataRevision, error) {
+	r := MetadataRevision(o)
+	if r < MetadataRevisionInitial {
+		return MetadataRevisionUninitialized,
+			fmt.Errorf("Cannot convert ordinal %s to a MetadataRevision", o)
+	}
+	return r, nil
+}
+
+func revisionToOrdinal(r MetadataRevision) (journalOrdinal, error) {
+	if r < MetadataRevisionInitial {
+		return journalOrdinal(0),
+			fmt.Errorf("Cannot convert revision %s to an ordinal", r)
+	}
+	return journalOrdinal(r), nil
+}
+
+// TODO: Consider caching the values returned by the read functions
+// below in memory.
+
+func (j mdServerBranchJournal) readEarliestRevision() (
+	MetadataRevision, error) {
+	o, err := j.j.readEarliestOrdinal()
+	if os.IsNotExist(err) {
+		return MetadataRevisionUninitialized, nil
+	} else if err != nil {
+		return MetadataRevisionUninitialized, err
+	}
+	return ordinalToRevision(o)
+}
+
+func (j mdServerBranchJournal) writeEarliestRevision(r MetadataRevision) error {
+	o, err := revisionToOrdinal(r)
+	if err != nil {
+		return err
+	}
+	return j.j.writeEarliestOrdinal(o)
+}
+
+func (j mdServerBranchJournal) readLatestRevision() (
+	MetadataRevision, error) {
+	o, err := j.j.readLatestOrdinal()
+	if os.IsNotExist(err) {
+		return MetadataRevisionUninitialized, nil
+	} else if err != nil {
+		return MetadataRevisionUninitialized, err
+	}
+	return ordinalToRevision(o)
+}
+
+func (j mdServerBranchJournal) writeLatestRevision(r MetadataRevision) error {
+	o, err := revisionToOrdinal(r)
+	if err != nil {
+		return err
+	}
+	return j.j.writeLatestOrdinal(o)
+}
+
+func (j mdServerBranchJournal) readMdID(r MetadataRevision) (MdID, error) {
+	o, err := revisionToOrdinal(r)
+	if err != nil {
+		return MdID{}, err
+	}
+	e, err := j.j.readJournalEntry(o)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	// TODO: Validate MdID?
+	return e.(MdID), nil
+}
+
+// All functions below are public functions.
+
+func (j mdServerBranchJournal) journalLength() (uint64, error) {
+	return j.j.journalLength()
+}
+
+func (j mdServerBranchJournal) getHead() (MdID, error) {
+	latestRevision, err := j.readLatestRevision()
+	if err != nil {
+		return MdID{}, err
+	} else if latestRevision == MetadataRevisionUninitialized {
+		return MdID{}, nil
+	}
+	return j.readMdID(latestRevision)
+}
+
+func (j mdServerBranchJournal) getRange(
+	start, stop MetadataRevision) (MetadataRevision, []MdID, error) {
+	earliestRevision, err := j.readEarliestRevision()
+	if err != nil {
+		return MetadataRevisionUninitialized, nil, err
+	} else if earliestRevision == MetadataRevisionUninitialized {
+		return MetadataRevisionUninitialized, nil, nil
+	}
+
+	latestRevision, err := j.readLatestRevision()
+	if err != nil {
+		return MetadataRevisionUninitialized, nil, err
+	} else if latestRevision == MetadataRevisionUninitialized {
+		return MetadataRevisionUninitialized, nil, nil
+	}
+
+	if start < earliestRevision {
+		start = earliestRevision
+	}
+
+	if stop > latestRevision {
```

Comment:
Hmm, I'd rather not return an error in that case, since that wouldn't match the behavior of `GetRange`. But added a conditional to return an uninitialized `earliestRevision` if the range is empty. (No caller relies on that behavior.)